templates/NDArray.kt

```diff
@@ -90,6 +90,53 @@ interface NDArray<T> {
      */
     fun toMutableList() = MutableList(size) { getGeneric(it) }
 
+    /**
+     * Determine whether every element of one NDArray is sufficiently close to the corresponding element
+     * of another array.  Two elements are considered close if
+     *
+     * abs(ele1 - ele2) < (atol + rtol * abs(ele1))
+     *
+     * @param other   the array to compare this one to
+     * @param rtol    the maximum relative (i.e. fractional) difference to allow between elements
+     * @param atol    the maximum absolute difference to allow between elements
+     */
+    fun <R> allClose(other: NDArray<R>, rtol: Double=1e-05, atol: Double=1e-08): Boolean {
+        if (!(shape().toIntArray() contentEquals other.shape().toIntArray()))
+            return false
+        for (i in 0 until size) {
+            val a = getDouble(i)
+            val b = other.getDouble(i)
+            val diff = kotlin.math.abs(a - b)
+            if (diff > atol + rtol*kotlin.math.abs(a))
+                return false
+        }
+        return true
+    }
+
+    /**
+     * Find the linear index of the minimum element in this array.
+     * If the array contains non-comparable values, this throws an exception.
+     */
+    fun argMin(): Int
```

Comment:
These could also be implemented as extension functions with a `T: Comparable` upper bound on their applicability. That is, something like `fun <T: Comparable<T>> NDArray<T>.argMin()`. Thus if the user has an `NDArray` holding something that is comparable `argMin` is a valid method, otherwise it's a compile time error.

Reply:
How about a hybrid approach, similar to how we implement [indexed access operators](https://github.com/kyonifer/koma/blob/master/koma-core-api/common/src/koma/extensions/extensions_matrix_Double.kt#L152)? The user would see an extension function `NDArray<Double>.argMin`, which would in turn delegate the action to a method on the instance with a different name like `argMinDouble` (for index accessors, the analog is a `NDArray<Double>.get(i)` extension function that forwards to `getDouble(i)`). 

The disadvantage is that right now the should-not-be-used-by-the-end-user-directly type-specific operations are still exposed to the user. There's a few potential ways this could be hidden from the user though:

- Marking the delegated methods as `protected`. This is unavailable since we are using an interface to implement `NDArray` instead of an abstract base (which would cause other issues).

- Marking the delegated methods as `internal`. This causes problems because the platform specific implementations are in a different project, which don't see internal declarations. However, this might be changing in the new multiplatform gradle plugin.

- Hiding them all on some internal class that the user likely won't dive into and could be overridden, e.g. declaring a `NumericalOperations` class that has all the platform specific stuff on it. I've considered doing this a few times to solve other problems in the past.

I think for now we could just do the straightforward hybrid approach with publically delegated functions like the array accessors do now, and then convert all of them together in the future if we ever decide to.

Reply:
Another perspective on this: BLAS level 1 operations like argmin are almost certainly memory bound and don't benefit from any sort of computational optimization a backend could bring. So I don't think there will be any immediate advantage to delegating these to the backends to do the work (my own benchmarks on the cblas k/native backend showed that implementing things like matrix addition directly in kotlin had no benefit over making the call into openblas). So maybe the right answer here is O(N) operations like finding the minimum we just go ahead and implement it once in the extension function.

The only concern there is that, for future GPU backends, such an implementation of `argmin` will require a recall of all the memory from the gpu to the cpu to do the work in kotlin. However, thats a bridge we can cross when we get there.

Reply:
> BLAS level 1 operations like argmin are almost certainly memory bound and don't benefit from any sort of computational optimization a backend could bring.

Actually there's a *lot* of room for optimization.  If the reduction is over the whole array, or the last axis, then memory access is fast.  You're accessing everything in order so the cache makes it efficient.  Otherwise you break the array up into tiles and compute several output values at once.  That way you still only have to load each cache line once.  Then you vectorize it and optionally multithread it.

Also, if accessing an individual element from Kotlin involves a native method call, putting the loop over elements in Kotlin rather than native code will be much slower.

Reply:
Not sure I'm following, but that may be due to my being too brief in reply again and therefore being unclear. I'm good with having a delegating virtual method on the NDArray instance that the extension function calls to leave our options open, so let's go with that.

Reply:
Ok, I think I just misunderstood.  So I'll do what you described in https://github.com/kyonifer/koma/pull/78#discussion_r226873257.  And for the moment I won't worry about trying to hide the internal method.

Reply:
Just to clarify, the base NDArray class will define all possible reductions for all types?  `argMinDouble()`, `argMinFloat()`, `argMinInt()`, etc., and the same for argMax, min, and max?  Then there will be an extension function `argMin()` on `NDArray<Double>` that calls `argMinDouble()`.  Is that right?

It's not obvious to me what advantage that has over just a single method for each one.  Perhaps I'm still misunderstanding.

Reply:
Just to clarify, the base NDArray class will define all possible reductions for all types?  `argMinDouble()`, `argMinFloat()`, `argMinInt()`, etc., and the same for argMax, min, and max?  Then there will be an extension function `argMin()` on `NDArray<Double>` that calls `argMinDouble()`.  Is that right?

It's not obvious to me what advantage that has over just a single method for each one.  Perhaps I'm still misunderstanding.

Reply:
Just to clarify, the base NDArray class will define all possible reductions for all types?  `argMinDouble()`, `argMinFloat()`, `argMinInt()`, etc., and the same for argMax, min, and max?  Then there will be an extension function `argMin()` on `NDArray<Double>` that calls `argMinDouble()`.  Is that right?

It's not obvious to me what advantage that has over just a single method for each one.  Perhaps I'm still misunderstanding.

Reply:
Sorry, I'm bad at explaining myself. Let me see if I can shed some light on the idea...

The goal of using extension functions instead of just having members on NDArray is type safety. Suppose we define:

```
fun <T: Comparable<T>> NDArray<T>.argMin(): Int {
  ...
}

```

then we would have

```
ndArrayOf(1, 2, 3) // Okay
ndArrayOf(1.0,2.0) // Okay
ndArrayOf(Image("a.jpg"), Image("b.jpg")).argMin() // Compile-time error, not Comparable
```
Thus trying to use `argMin` on something non-comparable becomes a syntax error at compile time. If `argMin` were defined on the instance itself (as it currently is in this PR) then `(foo as NDArray<Image>).argMin()` doesn't fail until runtime. This is noted in the current docstring: `If the array contains non-comparable values, this throws an exception.` It would be better to throw a compile time error than an exception, thus the recommendation to use extension functions.

So we could just end it there, and have the extension function contain a full implementation of `argMin`. However, as you noted there might be use cases where the backend wants to take control over that implementation. In this case we'll need to delegate the extension function implementation to virtual methods on `NDArray` that can be overridden by the implementation. Then we'd have:

```
fun <T: Comparable<T>> NDArray<T>.argMin(): Int {
    return this.argMinHiddenFromUser()
}
```

The virtual method should be "hidden" from the user one way or another, with the extension functions being what they see and call (the hiding part is TBD). Doing it this way with a user-facing extension function backed by a "hidden" virtual method on `NDArray` allows us to present the user with type safety and still delegate the implementation to the backends.

I don't think we need virtual methods `argMin$PRIMITIVE` for each primitive. We probably just need a single `argMinHiddenFromUser()` virtual method on `NDArray` which is overridden by the backends to do the operation in an efficient way for its particular data configuration, similar to what you currently have for `fun argMin(): Int` (except renamed to something hidden, so it doesn't override the extension function). The duplicate-methods-for-each-primitive thing is only needed when we are boxing on each individual element. For example, with the [array accessors](https://github.com/kyonifer/koma/blob/master/koma-core-api/common/src/koma/extensions/extensions_matrix_Double.kt#L152) we _did_ need separate methods for each primitive, because otherwise array access would be extremely slow (it was actually re-implemented this way after a user report that it was too slow to use). The decision as to whether or not we need one-per-primitive overloads of a method should hinge on whether or not we're incurring a box per element. For aggregate operations like finding the argmin over the entire array, there is no need. If there are methods that pass in/out elements one by one then we may need to look at whats getting boxed to decide.

Does that help clarify at all?